### Agent review — round 1

**Reviewer:** reviewer subagent (Kimi k2p5, via Local Operator `task`)
**Scope:** origin/main..HEAD (dbed0f1..fcf2ac9)

Focused tests: `.venv/bin/python -m pytest tests/unit/tools/test_eval_tool.py -q` → **41 passed**. All findings below were reproduced by driving the real `_handle` in this worktree, with the secret existing only in a file / env var — never in any cell source.

---

#### BLOCKER-1 — Chained exceptions leak argv: only the outermost exception is rewritten

`_format_error` calls `_redact_process_exception(exc)` on the outermost exception only (eval_worker.py:393), but `traceback.format_exception` renders the full `__context__`/`__cause__` chain. A `CalledProcessError` sitting in the chain keeps its real `cmd`.

Reproduction (secret lives only in a file; never disclosed):

```python
tok = open(dsn_file).read()          # 'TOPSECRET-FROMFILE-9x7'
try:
    subprocess.run(['false', 'db', tok], check=True)
except subprocess.CalledProcessError as e:
    raise RuntimeError(str(e))
```

The model-visible error contains the secret twice, verbatim:

```
subprocess.CalledProcessError: Command '['false', 'db', 'TOPSECRET-FROMFILE-9x7']' returned non-zero exit status 1.
...
RuntimeError: Command '['false', 'db', 'TOPSECRET-FROMFILE-9x7']' returned non-zero exit status 1.
LEAKED: True
```

This is not an exotic pattern — "catch low-level error, re-raise with context" is exactly what an agent does when a command fails. The same hole covers `raise X from e`, `e.add_note(str(cpe))`, and a hand-raised `CalledProcessError(1, ['cmd', tok])` where the re-raise detaches the original traceback (verified: `HANDRAISED LEAKED: True`). One-line fix direction: walk `__cause__`/`__context__` (cycle-guarded) rewriting each process exception's `cmd`, and/or make the backstop actually work (BLOCKER-2). Add a chain test — the current suite never exercises it, which is why this slipped through.

#### BLOCKER-2 — The message-level backstop never redacts anything

In `_redact_process_error_message` (eval_worker.py:159):

```python
redacted = [_redact_cmd_arg(t, [False]) for t in tokens]
```

A **fresh `[False]` is created per token**, so `seen_first` is always false for every token — every token is treated as argv[0] and returned verbatim. The function is a no-op on every input. Verified directly: even on its own target message shape, `AssertionError: ... Command '["make", "deploy", "AKIA-UNDISCLOSED"]' exited 2`, with an empty ledger, the count of changed tokens is **0** and the output is byte-identical. So the "backstop" the design relies on for chains/wrappers provides zero protection, and the "gated on the same ledger, so it never touches a command the model actually wrote" comment describes a function that touches nothing at all. Even with the `[False]` hoisted out of the comprehension, whole-traceback `shlex.split` tokenizes source frames and prose as "argv", so it needs a test proving it redacts a realistic wrapped message without mangling the frames (to its credit, I could not make it corrupt plain tracebacks — `AssertionError: don't give up` survives — precisely because it never fires).

#### MAJOR-1 — Unparseable string `cmd` passes through with the secret intact

`_redact_cmd` returns a string `cmd` unchanged when `shlex.split` raises `ValueError` (eval_worker.py:139-141). The comment calls this the safe direction, but it is the opposite: an undisclosed value inside an unsplittable string is rendered verbatim. Reproduced: `subprocess.run("curl -H 'Authorization: Bearer TOK endpoint", shell=True)` (unbalanced quote — one typo away from the normal form) renders the full string, token included, in the uncaught `CalledProcessError`. Fix direction: on parse failure, gate the **whole string** against the ledger — return it only if `discloses(cmd)`, else `<redacted:Nc>`.

#### MINOR-1 — "Byte-identical for all-disclosed commands" is false for string `cmd`

A disclosed string command with non-canonical spacing or quoting is normalized by the split/join: `_redact_cmd("echo  hi   there")` → `'echo hi there'`. Fidelity-only, and the list form (the common case) is unaffected, but the module comment (eval_worker.py:~93) and the docstring promise byte-identical passthrough. Fix the comment, or return `cmd` unchanged when every split token is disclosed.

---

**What checks out:** rewriting `exc.cmd` *is* reflected by `format_exception` (both `CalledProcessError` and `TimeoutExpired` read `self.cmd` at format time — verified); caught errors printing `e.cmd` still disclose explicitly (test passes, confirmed); bytes argv read from env is correctly redacted by the attribute rewrite (`<redacted:22c>`); circular `__cause__` chains don't hang the stdlib formatter; the ledger's substring-gating and eviction semantics are as documented, and over-redaction biases safe. The ledger design itself is sound — the holes are all in the rendering pass downstream of it.

**Verdict: NOT clean — 2 BLOCKERs, 1 MAJOR.** The direct uncaught-`run()` path (what all six new tests cover) works, but the chain/wrapper path leaks the exact credential class this PR exists to protect, and the backstop intended to cover that path is dead code. Another round is needed after the chain walk and backstop are fixed and covered by tests.

--- a/local_operator/tools/eval_worker.py
+++ b/local_operator/tools/eval_worker.py
@@ -245,6 +245,12 @@ def _redact_process_exception_chain(exc: BaseException) -> None:
         # __context__ is the implicit chain (an exception raised while handling
         # another); it is what a bare ``raise`` inside ``except`` produces.
         stack.append(current.__context__)
+        # format_exception renders ExceptionGroup LEAVES, so a CalledProcessError
+        # raised inside an asyncio.TaskGroup task is rendered too — and leaks
+        # argv if the walk stops at the group. Descend into .exceptions under
+        # the same visited set. BaseExceptionGroup carries the attribute; a
+        # plain exception simply does not, hence the getattr default.
+        stack.extend(getattr(current, "exceptions", ()))
 
 
 _REPR = reprlib.Repr()

--- a/local_operator/tools/eval_worker.py
+++ b/local_operator/tools/eval_worker.py
@@ -186,9 +186,11 @@ def _redact_cmd(cmd: Any) -> Any:
 
     A list is rebuilt element-wise; a string is split with ``shlex`` so a
     quoted argument stays one piece (quote awareness is fidelity, not safety:
-    without it ``sh -c 'exit 3'`` half-redacts an innocent line). Unparseable
-    input is returned unchanged — over-redaction of a string we cannot safely
-    split would mangle legitimate diagnostics.
+    without it ``sh -c 'exit 3'`` half-redacts an innocent line). A string
+    ``shlex`` cannot parse (one unbalanced quote from the normal case) is
+    ledger-gated whole: if its exact bytes appear in the model's source they
+    are already disclosed; otherwise it collapses to a single ``<redacted:Nc>``
+    rather than being rendered verbatim with the secret inside.
     """
     seen = [False]
     if isinstance(cmd, (list, tuple)):
@@ -197,34 +199,9 @@ def _redact_cmd(cmd: Any) -> Any:
         try:
             parts = shlex.split(cmd)
         except ValueError:
-            return cmd
+            return cmd if _ARGV_LEDGER.discloses(cmd) else _redacted_arg(cmd)
         return " ".join(_redact_cmd_arg(a, seen) for a in parts)
     return cmd
-
-
-def _redact_process_error_message(text: str) -> str:
-    """Redact an embedded ``Command '...'`` invocation in a rendered message.
-
-    The command line is rendered by the runtime inside the message, so we
-    recover it with the same ``shlex`` split and apply the disclosure gate
-    token by token. Only the first (argv[0]) token is preserved.
-    """
-    try:
-        tokens = shlex.split(text)
-    except ValueError:
-        return text
-    if not tokens:
-        return text
-    redacted = [_redact_cmd_arg(t, [False]) for t in tokens]
-    if redacted == tokens:
-        return text
-    # Replace longest-first so a redacted long token cannot be partially
-    # re-matched inside a shorter one it contains.
-    out = text
-    for original, masked in sorted(zip(tokens, redacted), key=lambda p: -len(p[0])):
-        if original != masked:
-            out = out.replace(original, masked)
-    return out
 
 
 def _redact_process_exception(exc: BaseException) -> None:
@@ -244,6 +221,30 @@ def _redact_process_exception(exc: BaseException) -> None:
         return
     with contextlib.suppress(Exception):
         exc.cmd = redacted  # type: ignore[attr-defined]
+
+
+def _redact_process_exception_chain(exc: BaseException) -> None:
+    """Redact argv in ``exc`` AND every process error in its context/cause chain.
+
+    ``format_exception`` renders the WHOLE ``__context__``/``__cause__`` chain,
+    so redacting only the outermost exception leaks a secret embedded in an
+    inner one — ``except CalledProcessError as e: raise RuntimeError(str(e))``
+    puts the undisclosed argv in the traceback twice. The walk is iterative
+    with a visited set so a cyclic chain (``raise ... from`` pointing back)
+    terminates instead of hanging.
+    """
+    visited: set[int] = set()
+    stack: list[BaseException | None] = [exc]
+    while stack:
+        current = stack.pop()
+        if current is None or id(current) in visited:
+            continue
+        visited.add(id(current))
+        _redact_process_exception(current)
+        stack.append(current.__cause__)
+        # __context__ is the implicit chain (an exception raised while handling
+        # another); it is what a bare ``raise`` inside ``except`` produces.
+        stack.append(current.__context__)
 
 
 _REPR = reprlib.Repr()
@@ -389,14 +390,10 @@ def _format_error(exc: BaseException) -> str:
     # exception's ``__str__`` interpolates ``cmd`` at format time, so this is
     # the single choke point. ``CalledProcessError.__str__`` reads ``self.cmd``
     # fresh, and ``TimeoutExpired.__str__`` likewise, so rewriting the
-    # attribute is what the formatted output reflects.
-    _redact_process_exception(exc)
-    rendered = "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)).strip()
-    # A chain (``raise ... from``) or a wrapper that rendered the command into
-    # its own message can still carry argv the attribute rewrite missed; the
-    # message-level pass is the backstop. It is gated on the same ledger, so it
-    # never touches a command the model actually wrote.
-    return _redact_process_error_message(rendered)
+    # attribute is what the formatted output reflects. The whole context/cause
+    # chain is walked because format_exception renders all of it.
+    _redact_process_exception_chain(exc)
+    return "".join(traceback.format_exception(type(exc), exc, exc.__traceback__)).strip()
 
 
 def _execute(namespace: dict[str, Any], code: str) -> str | None:

--- a/tests/unit/tools/test_eval_tool.py
+++ b/tests/unit/tools/test_eval_tool.py
@@ -561,6 +561,28 @@ def test_a_context_chain_also_redacts_the_inner_process_error(_fresh_argv_ledger
     assert "ValueError" in payload["error"] and "CalledProcessError" in payload["error"]
 
 
+def test_an_exception_group_leaf_also_redacts_argv(_fresh_argv_ledger, tmp_path) -> None:
+    """format_exception renders ExceptionGroup leaves: a CalledProcessError
+    raised inside an asyncio.TaskGroup task is part of the model-visible
+    traceback, so the walk must descend into .exceptions, not stop at the
+    group."""
+    secret = "taskgroup-secret"
+    sf = tmp_path / "s.txt"
+    sf.write_text(secret)
+    payload = _run_cell(
+        "import subprocess, asyncio\n"
+        f"tok = open({str(sf)!r}).read()\n"
+        "async def job():\n"
+        "    subprocess.run(['false', 'h', tok], check=True)\n"
+        "async def main():\n"
+        "    async with asyncio.TaskGroup() as tg:\n"
+        "        tg.create_task(job())\n"
+        "asyncio.run(main())\n"
+    )
+    assert "ExceptionGroup" in payload["error"]
+    assert secret not in payload["error"]
+
+
 def test_raise_from_redacts_the_cause_chain(_fresh_argv_ledger, tmp_path) -> None:
     """``raise ... from e`` names the cause explicitly; its argv is redacted."""
     secret = "raise-from-secret"

--- a/tests/unit/tools/test_eval_tool.py
+++ b/tests/unit/tools/test_eval_tool.py
@@ -541,6 +541,72 @@ def test_a_caught_error_is_an_explicit_disclosure_and_untouched(
     assert secret in payload["stdout"]
 
 
+def test_a_context_chain_also_redacts_the_inner_process_error(_fresh_argv_ledger, tmp_path) -> None:
+    """format_exception renders the WHOLE __context__/__cause__ chain, so
+    redacting only the outermost exception leaks the secret in the inner one.
+    A bare ``raise`` inside ``except`` is the common shape."""
+    secret = "context-chain-secret"
+    sf = tmp_path / "s.txt"
+    sf.write_text(secret)
+    payload = _run_cell(
+        "import subprocess\n"
+        f"tok = open({str(sf)!r}).read()\n"
+        "try:\n"
+        "    subprocess.run(['false', 'h', tok], check=True)\n"
+        "except subprocess.CalledProcessError:\n"
+        "    raise ValueError('wrapper')\n"
+    )
+    assert secret not in payload["error"]
+    # The wrapper and the inner error are both rendered, but the secret is not.
+    assert "ValueError" in payload["error"] and "CalledProcessError" in payload["error"]
+
+
+def test_raise_from_redacts_the_cause_chain(_fresh_argv_ledger, tmp_path) -> None:
+    """``raise ... from e`` names the cause explicitly; its argv is redacted."""
+    secret = "raise-from-secret"
+    sf = tmp_path / "s.txt"
+    sf.write_text(secret)
+    payload = _run_cell(
+        "import subprocess\n"
+        f"tok = open({str(sf)!r}).read()\n"
+        "try:\n"
+        "    subprocess.run(['false', 'h', tok], check=True)\n"
+        "except subprocess.CalledProcessError as e:\n"
+        "    raise KeyError('k') from e\n"
+    )
+    assert secret not in payload["error"]
+
+
+def test_an_unparseable_string_cmd_is_ledger_gated_not_passed_through(
+    _fresh_argv_ledger,
+) -> None:
+    """A string cmd shlex cannot parse (one unbalanced quote) must not be
+    rendered verbatim with the secret inside: the whole string is disclosed or
+    collapsed to a length marker."""
+    import subprocess as sp
+
+    from local_operator.tools import eval_worker
+
+    undisclosed = "curl -H 'Authorization: Bearer unbalanced-secret"
+    exc = sp.CalledProcessError(1, undisclosed)
+    eval_worker._redact_process_exception(exc)
+    assert exc.cmd == "<redacted:%dc>" % len(undisclosed)
+
+    # ...but a string the model literally wrote stays readable.
+    eval_worker._ARGV_LEDGER.record(undisclosed)
+    exc2 = sp.CalledProcessError(1, undisclosed)
+    eval_worker._redact_process_exception(exc2)
+    assert exc2.cmd == undisclosed
+
+
+def test_a_plain_traceback_without_a_command_is_untouched(_fresh_argv_ledger) -> None:
+    """The guard must not mangle ordinary errors. A ValueError whose message is
+    prose (no process invocation) renders word for word."""
+    payload = _run_cell("raise ValueError('plain words with spaces and no command')\n")
+    assert "plain words with spaces and no command" in payload["error"]
+    assert "<redacted" not in payload["error"]
+
+
 def test_worker_caps_stdout_stderr_display_and_repr_before_json() -> None:
     """Containment is worker-side: huge output never reaches the parent or
     json.dumps unbounded, even before the eval tool's 8KiB spill layer."""

